network/eni/geneve.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// +build linux
+
+package eni
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+
+	log "github.com/cihub/seelog"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+)
+
+type Geneve struct {
+	ENI
+	VNI               uint32
+	TunnelInterfaceIP net.IP
+	Dport             uint16
+	Primary           bool
+}
+
+// NewGeneve creates a new Geneve object.
+func NewGeneve(linkName string, vni string, tunnelIP net.IP, dPort uint16, primary bool) (*Geneve, error) {
+	if linkName == "" {
+		return nil, fmt.Errorf("Link name cannot be empty")
+	}
+	if vni == "" {
+		return nil, fmt.Errorf("VNI cannot be empty")
+	}
+	if tunnelIP == nil {
+		return nil, fmt.Errorf("tunnel interface IP cannot be empty")
+	}
+	if dPort == 0 {
+		return nil, fmt.Errorf("destination port number cannot be empty")
+	}
+
+	vniID, err := strconv.ParseInt(vni, 16, 32)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse VNI")
+	}
+
+	geneve := &Geneve{
+		ENI: ENI{
+			linkName: linkName,
+		},
+		VNI:               uint32(vniID),
+		TunnelInterfaceIP: tunnelIP,
+		Primary:           primary,
+		Dport:             dPort,
+	}
+
+	return geneve, nil
+}
+
+// AttachToLink attaches the Geneve object to an interface.
+func (geneve *Geneve) AttachToLink() error {
+	la := netlink.NewLinkAttrs()
+	la.Name = geneve.linkName
+
+	geneveLink := &netlink.Geneve{
+		LinkAttrs: la,
+		ID:        geneve.VNI,
+		Remote:    geneve.TunnelInterfaceIP,
+		Dport:     geneve.Dport,
+	}
+
+	log.Infof("Creating Geneve interface %s: %+v", geneve.linkName, geneveLink)
+	err := netlink.LinkAdd(geneveLink)
+	if err != nil {
+		if os.IsExist(err) {
+			log.Infof("Found existing Geneve interface %s.", geneve.linkName)
+		} else {
+			log.Errorf("Failed to add Geneve interface %s: %v", geneve.linkName, err)
+		}
```

Comment:
The caller logs these errors as well. Can't we just wrap the error and avoid logging here?

Reply:
Agreed that the `else` portion should be removed. Also `os.IsExist(err)` needs to be modified to `err.Error() == "device or resource busy"`. In case the resource is busy, we need to log it because the caller proceeds with other operations without logging the message. Will make the change in next revision.